### PR TITLE
SDK-232 [Node.js SDK] Add search_keys filter support

### PR DESCRIPTION
**Pull request by artemgolozubov**
Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- [SDK-232](https://incountry.atlassian.net/browse/SDK-232)

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [ ] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [x] Please put the target release vehicle this needs to be bundled with.
- [x] Did you increment the version in package.json? What is the new version?

### CODE REVIEW

File: src/validation/api/find-filter.ts [hunk 1]
```diff
@@ -39,7 +57,47 @@ const FilterNumberQueryIO: t.Type<FilterNumberQuery> = t.union([
...
+type FindFilterSearchField = {
+  [SEARCH_FIELD]: string;
+};
+
+const FindFilterSearchFieldValueIO: t.Type<string> = new t.Type(
+  'FindFilterSearchFieldValue',
+  (u): u is string => t.string.is(u) && u.length >= 3,
+  (u, c) => t.string.is(u) && u.length >= 3 ? t.success(u) : t.failure(u, c),
+  String,
+);
```
Comment (igor-alekseev-dev) [inline src/validation/api/find-filter.ts:71]: @artemgolozubov you can borrow complex helper to make validation of string length 
https://github.com/incountry/sdk-nodejs/pull/159/files#diff-9ed208834185e9249bdcaae6b57d2e03R79

**Review by RommeroInCountry — APPROVED** [review]
LGTM

File: README.md [hunk 2]
```diff
@@ -426,31 +426,66 @@ const readResult = await storage.read(countryCode, recordKey);
...
 - comparison operators for [Int fields](#int-fields-plain)
 ```typescript
+// Matches all records where record.rangeKey >= 5 AND record.rangeKey <= 100
 { rangeKey1: { $gte: 5, $lte: 100 } }
 ```
 
+##### Partial text match search
+
+Also you can search records by partial match using `searchKeys` operator, which performs partial match 
+search (using `LIKE` SQL operator) among record's text fields `key1, key2, ..., key10`.
```
Comment (RommeroInCountry) [inline README.md:471]: (using `LIKE` SQL operator) -> (similar to `LIKE` SQL operator)

File: README.md [hunk 3]
```diff
@@ -426,31 +426,66 @@ const readResult = await storage.read(countryCode, recordKey);
...
+```typescript
+// Matches all records where record.key1 LIKE 'abc' OR record.key2 LIKE 'abc' OR ... OR record.key10 LIKE 'abc'
+{ searchKeys: 'abc' }
+```
+
+**Please note:** `searchKeys` operator cannot be used in combination with any of `key1, key2, ..., key10` keys.
```
Comment (RommeroInCountry) [inline README.md:477]: and works only in combination with non-hashing Storage mode (hashSearchKeys param for Storage). 
We'll need a link to README's section about that

File: src/validation/api/find-filter.ts (hunk 1 above)
Comment (igor-alekseev-dev) [reply src/validation/api/find-filter.ts:71]: At least you can extract 
```
const is FindFilterSearchFieldValue = (u: unknown):  u is string => t.string.is(u) && u.length >= 3;

const FindFilterSearchFieldValueIO: t.Type<string> = new t.Type(
  'FindFilterSearchFieldValue',
  FindFilterSearchFieldValue,
  (u, c) => FindFilterSearchFieldValue(u) ? t.success(u) : t.failure(u, c),
  String,
);

```

File: src/validation/api/find-filter.ts [hunk 4]
```diff
@@ -39,7 +57,47 @@ const FilterNumberQueryIO: t.Type<FilterNumberQuery> = t.union([
...
+      }
+      if (intersection(Object.keys(u), EXCLUDED_KEYS_WHEN_SEARCHING).length > 0) {
+        return false;
+      }
+    }
+    return FindFilterWithoutSearchIO.is(u);
+  },
+  (u, c) => {
+    if (t.object.is(u) && Object.prototype.hasOwnProperty.call(u, SEARCH_FIELD)) {
+      if (isInvalid(FindFilterSearchFieldIO.validate(pick(u, SEARCH_FIELD), c))) {
```
Comment (igor-alekseev-dev) [inline src/validation/api/find-filter.ts:89]: `isInvalid(a.validate(b))` can be simplified to `!a.is(b)`

File: src/validation/api/find-filter.ts [hunk 5]
```diff
@@ -39,7 +57,47 @@ const FilterNumberQueryIO: t.Type<FilterNumberQuery> = t.union([
...
+  (u): u is string => t.string.is(u) && u.length >= 3,
+  (u, c) => t.string.is(u) && u.length >= 3 ? t.success(u) : t.failure(u, c),
+  String,
+);
+
+const FindFilterSearchFieldIO: t.Type<FindFilterSearchField> = t.record(t.literal(SEARCH_FIELD), FindFilterSearchFieldValueIO, 'FindFilterSearchField');
+const FindFilterIO = new t.Type(
+  'FindFilter',
+  (u): u is FindFilter => {
+    if (t.object.is(u) && Object.prototype.hasOwnProperty.call(u, SEARCH_FIELD)) {
```
Comment (igor-alekseev-dev) [inline src/validation/api/find-filter.ts:77]: Can be extracted to `isFindFilter` and used here and in validation

**Review by igor-alekseev-dev — DISMISSED** [review]

**Review by RommeroInCountry — APPROVED** [review]
LGTM

**Review by RommeroInCountry — APPROVED** [review]
LGTM

**Review by igor-alekseev-dev — APPROVED** [review]

**Review by vladimir-grezin — APPROVED** [review]
lgtm

**Pull request merged**